_transcripts/033-rust-1.50-1.51.md

```diff
@@ -0,0 +1,1232 @@
+single bit of data to describe. So if you have an option with `Some` contains a
+reference to anything, that `Option`, the size of it will only be the size of a
+single reference, and not the size of two, in fact, because you would actually,
+because of alignment and padding, or I guess just padding in this case, you
+would need to have double the size, right?
+
+__Jon__: Yeah, and I think it's interesting because it doesn't have to be just
+Option. And I think this is something that's been worked on internally, sort of
+in `rustc`, which is to find a way to— and this might have landed already on
+nightly, that the niche optimization is actually supported for, I think it's any
+enum where one variant does not hold data and there are only two variants.
+
+__Ben__: I think it's been it's been stable for a long, long time. It's not just
+nightly. I think this has been around for, like, years now that users have been
+able to take advantage of this.
+
+__Jon__: Yeah, but I think what's in nightly now is, if the niche is larger, and
+there are more variants, and all of them hold no data except for the primary
+one, then you can take advantage of that, too.
+
+__Ben__: Okay, yeah, it gets smarter all the time, so it's a pretty involved
+thing. I think it's smart enough to even have, like, if you have nested
+`Option`s, it might be able to contain the data for both variants inside of the
+niches.
+
+__Jon__: Yeah, depending on the size of the niche, right?
+
+__Ben__: Yeah.
+
+__Jon__: And I guess, to bring this back to the changelog, what has changed in
+in 1.50 is that a `File` on a Unix platform is represented as what's known as a
+file descriptor, which is really just a number. But it is specifically a signed
+number. And it's a number that cannot be `-1` because `-1` is used on Unix to
+represent that an operation failed and that you need to go check the error
+codes. And so `-1` is a niche for `File` on Unix platforms, because it just— a
+`File` cannot ever hold that value as its file descriptor. And so it can be used
+to represent `None` for an `Option<File>` instead.
+
+__Ben__: Yeah, squeezing out memory savings. That's a great use of Rust's time,
+I think.
+
+__Jon__: That's true.
+
+__Ben__: And those are all for the language features. We have some library
+stuff, unless you want to talk more about that.
+
+__Jon__: No, no, I think the library changes in 1.50 are are pretty interesting.
+The first one is, like, such an innocuous looking method. So this is the `then`
+method on `bool`. And the `then` method is pretty straightforward. If you have a
+`bool`, there's now a method on `bool` called `then`, that you pass in a
+closure, and then it returns an `Option` of the result of the closure. So the
+idea is that if the boolean is `false`, then `then` will not call the closure
+and return `None`. If the `bool` is `true`, then `then` will return `Some` of
+the value of calling the closure. So this is sort of a cheap, hacky way to turn
```

Comment:
Good prank played on the transcriber. Thank goodness for `code` formatting, or this would make no sense...
:)